core/commands/pin.go

```diff
@@ -200,25 +212,80 @@ Example:
 			res.SetError(err, cmds.ErrNormal)
 			return
 		}
-		if !found {
-			typeStr = "recursive"
-		}
 
-		switch typeStr {
-		case "all", "direct", "indirect", "recursive":
-		default:
-			err = fmt.Errorf("Invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
-			res.SetError(err, cmds.ErrClient)
+		if found {
+			switch typeStr {
+			case "all", "direct", "indirect", "recursive":
+			default:
+				err = fmt.Errorf("Invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
+				res.SetError(err, cmds.ErrClient)
+			}
 		}
 
+		argCount := len(req.Arguments())
+
 		keys := make(map[string]RefKeyObject)
-		if typeStr == "direct" || typeStr == "all" {
-			for _, k := range n.Pinning.DirectKeys() {
+
+		if (argCount > 0) {
+			if (!found) {
+				typeStr = "all"
+			}
+
+			for _, p := range req.Arguments() {
+				dagNode, err := core.Resolve(req.Context(), n, path.Path(p))
+
+				if err != nil {
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+
+				k, err := dagNode.Key()
+				if err != nil {
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+
+				pinType, pinned, err := n.Pinning.IsPinnedWithType(k, typeStr)
+				if err != nil {
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+
+				if !pinned {
+					err = fmt.Errorf("Path '%s' is not pinned", p)
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+
+				switch pinType {
+				case "direct", "indirect", "recursive", "internal":
+				default:
```

Comment:
In isPinnedWithType() in pin/pin.go around line 214, in case of indirect pin there is:

```
	// Default is "indirect"
	for _, rk := range p.recursePin.GetKeys() {
		rnd, err := p.dserv.Get(context.Background(), rk)
		if err != nil {
			return "", false, err
		}

		has, err := hasChild(p.dserv, rnd, k)
		if err != nil {
			return "", false, err
		}
		if has {
			return rk.B58String(), true, nil
		}
	}
	return "", false, nil
```

So when the pin is indirect it's not "indirect" that is returned but the hash of the recursively pinned object that is responsible for the current object being pinned.